src/metabase/driver/druid.clj

```diff
@@ -1,13 +1,213 @@
 (ns metabase.driver.druid
   "Druid driver."
-  (:require [clojure.tools.logging :as log]
+  (:require
+
+[clojure.math.numeric-tower :as math]
+            [clojure.string :as s]
+
+[clojure.tools.logging :as log]
             [clj-http.client :as http]
             [cheshire.core :as json]
             [metabase.driver :as driver]
             [metabase.driver.druid.query-processor :as qp]
             (metabase.models [field :as field]
                              [table :as table])
-            [metabase.util :as u]))
+            [metabase.sync-database.analyze :as analyze]
+            [metabase.db.metadata-queries :as queries]
+            [metabase.util :as u]
+
+            [metabase.query-processor :as qp1]
+            [metabase.query-processor.expand :as ql]
+))
+
+;; Analyze routines begin
+
+
+(defn- qp-query [db-id query]
+               (binding [qp1/*disable-qp-logging* false]
+  (-> (qp1/process-query
+       {:type     :query
+        :database db-id
+        :query    query})
+      :data
+      :rows)))
+
+
+(def ^:private ^:const low-cardinality-threshold
+  "Fields with less than this many distinct values should automatically be marked with `special_type = :category`."
+  40)
+
+(def ^:private ^:const field-values-entry-max-length
+  "The maximum character length for a stored `FieldValues` entry."
+  100)
+
+(defn- field-query [field query]
+  (let [table (field/table field)]
+    (qp-query (:db_id table)
+              (ql/query (merge query)
+                        (ql/source-table (:id table))))))
+
+(defn druid-field-distinct-values
+  "Return the distinct values of FIELD.
+   This is used to create a `FieldValues` object for `:category` Fields."
+  ([field]
+    (druid-field-distinct-values field @(resolve low-cardinality-threshold)))
+  ([{field-id :id :as field} max-results]
+   {:pre [(integer? max-results)]}
+   (mapv first (field-query field (-> {}
+                                      (ql/breakout (ql/field-id field-id))
+                                      (ql/limit max-results))))
+   #_(let [table (field/table field)
+db-id (:db_id table)
+
+q1 {:query
+ {:intervals ["1900-01-01/2100-01-01"],
+  :granularity :all,
+  :context {:timeout 60000},
+  :queryType :topN,
+  :threshold 2000,
+  :dataSource "obi-billdata-qa-1",
+  :aggregations [{:type :count, :name :___count}],
+  :dimension "payment_method",
+  :metric {:type :alphaNumeric}},
+ :querytype :metabase.driver.druid.query-processor/topN}
+
+
+]
+(binding [metabase.query-processor/*disable-qp-logging* false]
+ (metabase.query-processor/process-query (merge {
+:type :query
+:database db-id}
+q1))
+
+)
+
+     )
+   ))
```

Comment:
having a hard time trying to understand what's going on here, this is not how Clojure code is usually indented... 😓